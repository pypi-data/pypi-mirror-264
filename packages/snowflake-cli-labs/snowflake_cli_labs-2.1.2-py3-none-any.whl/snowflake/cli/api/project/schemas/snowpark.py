from __future__ import annotations

from snowflake.cli.api.project.schemas.relaxed_map import RelaxedMap
from snowflake.cli.api.project.util import IDENTIFIER
from strictyaml import (
    Bool,
    EmptyList,
    MapPattern,
    Optional,
    Regex,
    Seq,
    Str,
)

Argument = RelaxedMap({"name": Str(), "type": Str(), Optional("default"): Str()})

_callable_mapping = {
    "name": Str(),
    Optional("database", default=None): Regex(IDENTIFIER),
    Optional("schema", default=None): Regex(IDENTIFIER),
    "handler": Str(),
    "returns": Str(),
    "signature": Seq(Argument) | EmptyList(),
    Optional("runtime"): Str(),
    Optional("external_access_integration"): Seq(Str()),
    Optional("secrets"): MapPattern(Str(), Str()),
    Optional("imports"): Seq(Str()),
}

function_schema = RelaxedMap(_callable_mapping)

procedure_schema = RelaxedMap(
    {
        **_callable_mapping,
        Optional("execute_as_caller"): Bool(),
    }
)

snowpark_schema = RelaxedMap(
    {
        "project_name": Str(),
        "stage_name": Str(),
        "src": Str(),
        Optional("functions"): Seq(function_schema),
        Optional("procedures"): Seq(procedure_schema),
    }
)
