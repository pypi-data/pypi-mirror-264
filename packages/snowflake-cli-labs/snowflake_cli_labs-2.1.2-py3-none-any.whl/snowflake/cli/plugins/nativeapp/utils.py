from sys import stdin, stdout


def needs_confirmation(needs_confirm: bool, auto_yes: bool) -> bool:
    return needs_confirm and not auto_yes


def is_tty_interactive():
    return stdin.isatty() and stdout.isatty()
