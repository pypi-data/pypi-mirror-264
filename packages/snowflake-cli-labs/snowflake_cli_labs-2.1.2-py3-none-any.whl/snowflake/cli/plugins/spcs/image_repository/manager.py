from urllib.parse import urlparse

from click import ClickException
from snowflake.cli.api.constants import ObjectType
from snowflake.cli.api.project.util import (
    is_valid_unquoted_identifier,
)
from snowflake.cli.api.sql_execution import SqlExecutionMixin
from snowflake.cli.plugins.spcs.common import handle_object_already_exists
from snowflake.connector.errors import ProgrammingError


class ImageRepositoryManager(SqlExecutionMixin):
    def get_database(self):
        return self._conn.database

    def get_schema(self):
        return self._conn.schema

    def get_role(self):
        return self._conn.role

    def get_repository_url(self, repo_name: str, with_scheme: bool = True):
        if not is_valid_unquoted_identifier(repo_name):
            raise ValueError(
                f"repo_name '{repo_name}' is not a valid unquoted Snowflake identifier"
            )
        # we explicitly do not allow this function to be used without connection database and schema set
        repo_row = self.show_specific_object(
            "image repositories", repo_name, check_schema=True
        )
        if repo_row is None:
            raise ClickException(
                f"Image repository '{repo_name}' does not exist in database '{self.get_database()}' and schema '{self.get_schema()}' or not authorized."
            )
        if with_scheme:
            return f"https://{repo_row['repository_url']}"
        else:
            return repo_row["repository_url"]

    def get_repository_api_url(self, repo_url):
        """
        Converts a repo URL to a registry OCI API URL.
        https://reg.com/db/schema/repo becomes https://reg.com/v2/db/schema/repo
        """
        parsed_url = urlparse(repo_url)

        scheme = parsed_url.scheme
        host = parsed_url.netloc
        path = parsed_url.path

        return f"{scheme}://{host}/v2{path}"

    def create(self, name: str):
        try:
            return self._execute_schema_query(f"create image repository {name}")
        except ProgrammingError as e:
            handle_object_already_exists(e, ObjectType.IMAGE_REPOSITORY, name)
