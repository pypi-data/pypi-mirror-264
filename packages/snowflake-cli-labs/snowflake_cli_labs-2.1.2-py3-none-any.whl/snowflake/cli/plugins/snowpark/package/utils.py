from __future__ import annotations

from dataclasses import dataclass
from pathlib import Path
from typing import List

from requirements.requirement import Requirement
from snowflake.cli.api.secure_path import SecurePath
from snowflake.cli.plugins.snowpark.models import SplitRequirements


@dataclass
class LookupResult:
    requirements: SplitRequirements
    name: str

    @property
    def message(self):
        return ""


class InAnaconda(LookupResult):
    @property
    def message(self):
        return f"Package {self.name} is available on the Snowflake Anaconda channel."


class RequiresPackages(LookupResult):
    @property
    def message(self):
        return f"""The package {self.name} is supported, but does depend on the
                following Snowflake supported libraries. You should
                include the following in your packages:
                {get_readable_list_of_requirements(self.requirements.snowflake)}"""


class NotInAnaconda(LookupResult):
    @property
    def message(self):
        return f"""The package {self.name} is avaiable through PIP. You can create a zip using:\n
                snow snowpark package create {self.name} --pypi-download"""


class NothingFound(LookupResult):
    @property
    def message(self):
        return f"""Nothing found for {self.name}. Most probably, package is not avaiable on Snowflake Anaconda channel\n
                    Please check the package name or try again with --pypi-download option"""


@dataclass
class CreateResult:
    package_name: str
    file_name: Path = Path()


class CreatedSuccessfully(CreateResult):
    @property
    def message(self):
        return f"Package {self.package_name}.zip created. You can now upload it to a stage (`snow snowpark package upload -f {self.package_name}.zip -s <stage-name>`) and reference it in your procedure or function."


def prepare_app_zip(file_path: SecurePath, temp_dir: SecurePath) -> SecurePath:
    # get filename from file path (e.g. app.zip from /path/to/app.zip)
    # TODO: think if no file exceptions are handled correctly
    file_name = file_path.path.name
    temp_path = temp_dir / file_name
    file_path.copy(temp_path.path)
    return temp_path


def get_readable_list_of_requirements(reqs: List[Requirement]):
    return "\n".join((req.line for req in reqs))
