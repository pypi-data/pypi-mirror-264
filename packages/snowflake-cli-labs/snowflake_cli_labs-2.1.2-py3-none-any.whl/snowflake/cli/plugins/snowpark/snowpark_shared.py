from __future__ import annotations

import logging
from pathlib import Path
from typing import List

import click
import typer
from requirements.requirement import Requirement
from snowflake.cli.api.secure_path import SecurePath
from snowflake.cli.plugins.snowpark import package_utils
from snowflake.cli.plugins.snowpark.models import PypiOption
from snowflake.cli.plugins.snowpark.zipper import zip_dir

PyPiDownloadOption: PypiOption = typer.Option(
    PypiOption.ASK.value, help="Whether to download non-Anaconda packages from PyPi."
)

PackageNativeLibrariesOption: PypiOption = typer.Option(
    PypiOption.NO.value,
    help="Allows native libraries, when using packages installed through PIP",
)

CheckAnacondaForPyPiDependencies: bool = typer.Option(
    True,
    "--check-anaconda-for-pypi-deps/--no-check-anaconda-for-pypi-deps",
    "-a",
    help="""Checks if any of missing Anaconda packages dependencies can be imported directly from Anaconda. Valid values include: `true`, `false`, Default: `true`.""",
)

ReturnsOption = typer.Option(
    ...,
    "--returns",
    "-r",
    help="Data type for the procedure to return.",
)

OverwriteOption = typer.Option(
    False,
    "--overwrite",
    "-o",
    help="Replaces an existing procedure with this one.",
)

log = logging.getLogger(__name__)

REQUIREMENTS_SNOWFLAKE = "requirements.snowflake.txt"
REQUIREMENTS_OTHER = "requirements.other.txt"


def snowpark_package(
    source: Path,
    artifact_file: Path,
    pypi_download: PypiOption,
    check_anaconda_for_pypi_deps: bool,
    package_native_libraries: PypiOption,
):
    log.info("Resolving any requirements from requirements.txt...")
    requirements = package_utils.parse_requirements()
    if requirements:
        log.info("Comparing provided packages from Snowflake Anaconda...")
        split_requirements = package_utils.parse_anaconda_packages(requirements)
        if not split_requirements.other:
            log.info("No packages to manually resolve")
        else:
            _write_requirements_file(REQUIREMENTS_OTHER, split_requirements.other)
            do_download = (
                click.confirm(
                    "Do you want to try to download non-Anaconda packages?",
                    default=True,
                )
                if pypi_download == PypiOption.ASK
                else pypi_download == PypiOption.YES
            )
            if do_download:
                log.info("Installing non-Anaconda packages...")
                should_continue, second_chance_results = package_utils.install_packages(
                    REQUIREMENTS_OTHER,
                    check_anaconda_for_pypi_deps,
                    package_native_libraries,
                )
                # add the Anaconda packages discovered as dependencies
                if should_continue and second_chance_results:
                    split_requirements.snowflake = (
                        split_requirements.snowflake + second_chance_results.snowflake
                    )

        # write requirements.snowflake.txt file
        if split_requirements.snowflake:
            _write_requirements_file(
                REQUIREMENTS_SNOWFLAKE,
                package_utils.deduplicate_and_sort_reqs(split_requirements.snowflake),
            )

    zip_dir(source=source, dest_zip=artifact_file)

    if Path(".packages").exists():
        zip_dir(source=Path(".packages"), dest_zip=artifact_file, mode="a")
    log.info("Deployment package now ready: %s", artifact_file)


def _write_requirements_file(file_name: str, requirements: List[Requirement]):
    log.info("Writing %s file", file_name)
    with SecurePath(file_name).open("w", encoding="utf-8") as f:
        for req in requirements:
            f.write(f"{req.line}\n")
