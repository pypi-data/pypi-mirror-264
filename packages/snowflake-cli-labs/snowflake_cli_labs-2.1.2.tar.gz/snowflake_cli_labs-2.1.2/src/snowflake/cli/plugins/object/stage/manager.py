from __future__ import annotations

import logging
import re
from contextlib import nullcontext
from pathlib import Path
from typing import Optional, Union

from snowflake.cli.api.project.util import to_string_literal
from snowflake.cli.api.sql_execution import SqlExecutionMixin
from snowflake.cli.api.utils.path_utils import path_resolver
from snowflake.connector.cursor import SnowflakeCursor

log = logging.getLogger(__name__)


UNQUOTED_FILE_URI_REGEX = r"[\w/*?\-.=&{}$#[\]\"\\!@%^+:]+"


class StageManager(SqlExecutionMixin):
    @staticmethod
    def get_standard_stage_name(name: str) -> str:
        # Handle embedded stages
        if name.startswith("snow://") or name.startswith("@"):
            return name

        return f"@{name}"

    @staticmethod
    def get_stage_name_from_path(path: str):
        """
        Returns stage name from potential path on stage. For example
        db.schema.stage/foo/bar  -> db.schema.stage
        """
        return Path(path).parts[0]

    @staticmethod
    def quote_stage_name(name: str) -> str:
        if name.startswith("'") and name.endswith("'"):
            return name  # already quoted

        standard_name = StageManager.get_standard_stage_name(name)
        if standard_name.startswith("@") and not re.fullmatch(
            r"@([\w./$])+", standard_name
        ):
            return to_string_literal(standard_name)

        return standard_name

    def _to_uri(self, local_path: str):
        uri = f"file://{local_path}"
        if re.fullmatch(UNQUOTED_FILE_URI_REGEX, uri):
            return uri
        return to_string_literal(uri)

    def list_files(self, stage_name: str) -> SnowflakeCursor:
        stage_name = self.get_standard_stage_name(stage_name)
        return self._execute_query(f"ls {self.quote_stage_name(stage_name)}")

    def get(
        self, stage_name: str, dest_path: Path, parallel: int = 4
    ) -> SnowflakeCursor:
        stage_name = self.get_standard_stage_name(stage_name)
        dest_directory = f"{dest_path}/"
        return self._execute_query(
            f"get {self.quote_stage_name(stage_name)} {self._to_uri(dest_directory)} parallel={parallel}"
        )

    def put(
        self,
        local_path: Union[str, Path],
        stage_path: str,
        parallel: int = 4,
        overwrite: bool = False,
        role: Optional[str] = None,
    ) -> SnowflakeCursor:
        """
        This method will take a file path from the user's system and put it into a Snowflake stage,
        which includes its fully qualified name as well as the path within the stage.
        If provided with a role, then temporarily use this role to perform the operation above,
        and switch back to the original role for the next commands to run.
        """
        with self.use_role(role) if role else nullcontext():
            stage_path = self.get_standard_stage_name(stage_path)
            local_resolved_path = path_resolver(str(local_path))
            log.info("Uploading %s to @%s", local_resolved_path, stage_path)
            cursor = self._execute_query(
                f"put {self._to_uri(local_resolved_path)} {self.quote_stage_name(stage_path)} "
                f"auto_compress=false parallel={parallel} overwrite={overwrite}"
            )
        return cursor

    def remove(
        self, stage_name: str, path: str, role: Optional[str] = None
    ) -> SnowflakeCursor:
        """
        This method will take a file path that exists on a Snowflake stage,
        and remove it from the stage.
        If provided with a role, then temporarily use this role to perform the operation above,
        and switch back to the original role for the next commands to run.
        """
        with self.use_role(role) if role else nullcontext():
            stage_name = self.get_standard_stage_name(stage_name)
            path = path if path.startswith("/") else "/" + path
            quoted_stage_name = self.quote_stage_name(f"{stage_name}{path}")
            return self._execute_query(f"remove {quoted_stage_name}")

    def create(self, stage_name: str, comment: Optional[str] = None) -> SnowflakeCursor:
        query = f"create stage if not exists {stage_name}"
        if comment:
            query += f" comment='{comment}'"
        return self._execute_query(query)
