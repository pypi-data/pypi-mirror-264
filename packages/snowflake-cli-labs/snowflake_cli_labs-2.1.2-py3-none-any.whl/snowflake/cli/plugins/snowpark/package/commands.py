from __future__ import annotations

import logging
from pathlib import Path

import typer
from snowflake.cli.api.commands.snow_typer import SnowTyper
from snowflake.cli.api.output.types import CommandResult, MessageResult
from snowflake.cli.plugins.snowpark.package.manager import (
    cleanup_after_install,
    create,
    lookup,
    upload,
)
from snowflake.cli.plugins.snowpark.package.utils import (
    CreatedSuccessfully,
    NotInAnaconda,
    RequiresPackages,
)

app = SnowTyper(
    name="package",
    help="Manages custom Python packages for Snowpark",
)
log = logging.getLogger(__name__)

install_option = typer.Option(
    False,
    "--pypi-download",
    help="Installs packages that are not available on the Snowflake Anaconda channel.",
)

deprecated_install_option = typer.Option(
    False,
    "--yes",
    "-y",
    hidden=True,
    help="Installs packages that are not available on the Snowflake Anaconda channel.",
)


@app.command("lookup", requires_connection=True)
def package_lookup(
    name: str = typer.Argument(..., help="Name of the package."),
    install_packages: bool = install_option,
    deprecated_install_option: bool = deprecated_install_option,
    **options,
) -> CommandResult:
    """
    Checks if a package is available on the Snowflake Anaconda channel.
    If the `--pypi-download` flag is provided, this command checks all dependencies of the packages
    outside Snowflake channel.
    """
    if deprecated_install_option:
        install_packages = deprecated_install_option

    lookup_result = lookup(name=name, install_packages=install_packages)
    cleanup_after_install()
    return MessageResult(lookup_result.message)


@app.command("upload", requires_connection=True)
def package_upload(
    file: Path = typer.Option(
        ...,
        "--file",
        "-f",
        help="Path to the file to upload.",
        exists=False,
    ),
    stage: str = typer.Option(
        ...,
        "--stage",
        "-s",
        help="Name of the stage in which to upload the file, not including the @ symbol.",
    ),
    overwrite: bool = typer.Option(
        False,
        "--overwrite",
        "-o",
        help="Overwrites the file if it already exists.",
    ),
    **options,
) -> CommandResult:
    """
    Uploads a Python package zip file to a Snowflake stage so it can be referenced in the imports of a procedure or function.
    """
    return MessageResult(upload(file=file, stage=stage, overwrite=overwrite))


@app.command("create", requires_connection=True)
def package_create(
    name: str = typer.Argument(
        ...,
        help="Name of the package to create.",
    ),
    install_packages: bool = install_option,
    deprecated_install_option: bool = deprecated_install_option,
    **options,
) -> CommandResult:
    """
    Creates a Python package as a zip file that can be uploaded to a stage and imported for a Snowpark Python app.
    """
    if deprecated_install_option:
        install_packages = deprecated_install_option

    if (
        type(lookup_result := lookup(name=name, install_packages=install_packages))
        in [
            NotInAnaconda,
            RequiresPackages,
        ]
        and type(creation_result := create(name)) == CreatedSuccessfully
    ):
        message = creation_result.message
        if type(lookup_result) == RequiresPackages:
            message += "\n" + lookup_result.message
    else:
        message = lookup_result.message

    cleanup_after_install()
    return MessageResult(message)
