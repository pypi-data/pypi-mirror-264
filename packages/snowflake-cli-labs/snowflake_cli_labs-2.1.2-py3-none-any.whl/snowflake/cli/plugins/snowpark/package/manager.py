from __future__ import annotations

import logging
import os.path
from pathlib import Path

from requirements.requirement import Requirement
from snowflake.cli.api.constants import PACKAGES_DIR
from snowflake.cli.api.secure_path import SecurePath
from snowflake.cli.plugins.object.stage.manager import StageManager
from snowflake.cli.plugins.snowpark import package_utils
from snowflake.cli.plugins.snowpark.models import SplitRequirements
from snowflake.cli.plugins.snowpark.package.utils import (
    CreatedSuccessfully,
    InAnaconda,
    LookupResult,
    NothingFound,
    NotInAnaconda,
    RequiresPackages,
    prepare_app_zip,
)
from snowflake.cli.plugins.snowpark.zipper import zip_dir

log = logging.getLogger(__name__)


def lookup(name: str, install_packages: bool) -> LookupResult:

    package_response = package_utils.parse_anaconda_packages([Requirement.parse(name)])

    if package_response.snowflake and not package_response.other:
        return InAnaconda(package_response, name)
    elif install_packages:
        status, result = package_utils.install_packages(
            perform_anaconda_check=True, package_name=name, file_name=None
        )

        if status:
            if result.snowflake:
                return RequiresPackages(result, name)
            else:
                return NotInAnaconda(result, name)

    return NothingFound(SplitRequirements([], []), name)


def upload(file: Path, stage: str, overwrite: bool):
    log.info("Uploading %s to Snowflake @%s/%s...", file, stage, file)
    with SecurePath.temporary_directory() as temp_dir:
        temp_app_zip_path = prepare_app_zip(SecurePath(file), temp_dir)
        sm = StageManager()

        sm.create(sm.get_stage_name_from_path(stage))
        put_response = sm.put(
            temp_app_zip_path.path, stage, overwrite=overwrite
        ).fetchone()

    message = f"Package {file} {put_response[6]} to Snowflake @{stage}/{file}."

    if put_response[6] == "SKIPPED":
        message = "Package already exists on stage. Consider using --overwrite to overwrite the file."

    return message


def create(zip_name: str):
    file_name = zip_name if zip_name.endswith(".zip") else f"{zip_name}.zip"
    zip_dir(dest_zip=Path(file_name), source=Path.cwd() / ".packages")

    if os.path.exists(file_name):
        return CreatedSuccessfully(zip_name, Path(file_name))


def cleanup_after_install():
    if PACKAGES_DIR.exists():
        SecurePath(PACKAGES_DIR).rmdir(recursive=True)
