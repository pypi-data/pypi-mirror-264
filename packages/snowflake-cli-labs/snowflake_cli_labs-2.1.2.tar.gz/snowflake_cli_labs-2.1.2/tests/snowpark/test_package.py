import logging
import os
from unittest import mock
from unittest.mock import MagicMock, patch
from zipfile import ZipFile

import pytest
from requirements.requirement import Requirement

import snowflake.cli.plugins.snowpark.package.manager
import src.snowflake.cli.plugins.snowpark.package.manager
from snowflake.cli.plugins.snowpark.models import SplitRequirements
from snowflake.cli.plugins.snowpark.package.utils import NotInAnaconda, NothingFound

from tests.test_data import test_data


class TestPackage:
    @pytest.mark.parametrize(
        "argument",
        ["snowflake-connector-python", "some-weird-package-we-dont-know"],
    )
    @patch("snowflake.cli.plugins.snowpark.package_utils.requests")
    def test_package_lookup(
        self, mock_requests, argument, monkeypatch, runner, snapshot
    ) -> None:
        mock_requests.get.return_value = self.mocked_anaconda_response(
            test_data.anaconda_response
        )

        result = runner.invoke(["snowpark", "package", "lookup", argument, "--yes"])

        assert result.exit_code == 0
        assert result.output == snapshot

    @patch("snowflake.cli.plugins.snowpark.package_utils.install_packages")
    @patch("snowflake.cli.plugins.snowpark.package_utils.parse_anaconda_packages")
    def test_package_lookup_with_install_packages(
        self, mock_package, mock_install, runner, capfd, snapshot
    ) -> None:

        mock_package.return_value = SplitRequirements(
            [], [Requirement("some-other-package")]
        )
        mock_install.return_value = (
            True,
            SplitRequirements(
                [Requirement("snowflake-snowpark-python")],
                [Requirement("some-other-package")],
            ),
        )

        result = runner.invoke(
            ["snowpark", "package", "lookup", "some-other-package", "--yes"]
        )
        assert result.exit_code == 0
        assert result.output == snapshot

    @patch("snowflake.cli.plugins.snowpark.package.commands.lookup")
    def test_package_create(
        self, mock_lookup, caplog, temp_dir, dot_packages_directory, runner
    ) -> None:

        mock_lookup.return_value = NotInAnaconda(
            SplitRequirements([], ["some-other-package"]), "totally-awesome-package"
        )

        with caplog.at_level(
            logging.DEBUG, logger="snowflake.cli.plugins.snowpark.package"
        ):
            result = runner.invoke(
                ["snowpark", "package", "create", "totally-awesome-package", "--yes"]
            )

        assert result.exit_code == 0
        assert os.path.isfile("totally-awesome-package.zip")

        zip_file = ZipFile("totally-awesome-package.zip", "r")

        assert (
            "totally-awesome-package/totally-awesome-module.py" in zip_file.namelist()
        )
        os.remove("totally-awesome-package.zip")

    @mock.patch("snowflake.cli.plugins.snowpark.package.manager.StageManager")
    @mock.patch("snowflake.connector.connect")
    def test_package_upload(
        self,
        mock_connector,
        mock_stage_manager,
        package_file: str,
        runner,
        mock_ctx,
        mock_cursor,
    ) -> None:
        ctx = mock_ctx()
        mock_connector.return_value = ctx
        mock_stage_manager().put.return_value = mock_cursor(
            rows=[("", "", "", "", "", "", "UPLOADED")], columns=[]
        )

        result = runner.invoke(
            ["snowpark", "package", "upload", "-f", package_file, "-s", "stageName"]
        )

        assert result.exit_code == 0
        assert ctx.get_query() == ""

    @mock.patch(
        "snowflake.cli.plugins.snowpark.package.manager.StageManager._execute_query"
    )
    def test_package_upload_to_path(
        self,
        mock_execute_queries,
        package_file: str,
        runner,
        mock_ctx,
        mock_cursor,
    ) -> None:
        mock_execute_queries.return_value = MagicMock()

        result = runner.invoke(
            [
                "snowpark",
                "package",
                "upload",
                "-f",
                package_file,
                "-s",
                "db.schema.stage/path/to/file",
            ]
        )

        assert result.exit_code == 0
        assert mock_execute_queries.call_count == 2
        create, put = mock_execute_queries.call_args_list
        assert create.args[0] == "create stage if not exists db.schema.stage"
        assert "db.schema.stage/path/to/file" in put.args[0]

    @pytest.mark.parametrize("command", ["lookup", "create"])
    @pytest.mark.parametrize(
        "flags,expected_value",
        [
            (["--pypi-download"], True),
            (["-y"], True),
            (["--yes"], True),
            (["--pypi-download", "-y"], True),
            ([], False),
        ],
    )
    @mock.patch("snowflake.cli.plugins.snowpark.package.commands.lookup")
    def test_install_flag(self, mock_lookup, command, flags, expected_value, runner):
        mock_lookup.return_value = NothingFound
        result = runner.invoke(["snowpark", "package", "lookup", "foo", *flags])

        mock_lookup.assert_called_with(name="foo", install_packages=expected_value)

    @staticmethod
    def mocked_anaconda_response(response: dict):
        mock_response = MagicMock()
        mock_response.status_code = 200
        mock_response.json.return_value = response

        return mock_response
