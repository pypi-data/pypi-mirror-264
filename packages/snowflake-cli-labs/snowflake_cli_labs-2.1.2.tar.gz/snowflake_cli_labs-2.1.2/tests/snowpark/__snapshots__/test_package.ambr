# serializer version: 1
# name: TestPackage.test_package_lookup[snowflake-connector-python]
  '''
  Package snowflake-connector-python is available on the Snowflake Anaconda channel.
  
  '''
# ---
# name: TestPackage.test_package_lookup[some-weird-package-we-dont-know]
  '''
  Nothing found for some-weird-package-we-dont-know. Most probably, package is not avaiable on Snowflake Anaconda channel
  
                      Please check the package name or try again with --pypi-download option
  
  '''
# ---
# name: TestPackage.test_package_lookup_with_install_packages
  '''
  The package some-other-package is supported, but does depend on the
                  following Snowflake supported libraries. You should
                  include the following in your packages:
                  snowflake-snowpark-python
  
  '''
# ---
